src/GraphQL/Types/Composite/InputObjectGraphType.cs

```diff
@@ -50,12 +52,46 @@ internal InputObjectGraphType(InputObjectGraphType<TSourceType>? cloneFrom)
             : base(cloneFrom)
         {
             // if (cloneFrom == null) { /* initialization logic */ }
+
+            if (typeof(TSourceType) == typeof(object))
+            {
+                // for InputObjectGraphType just return the dictionary
+                _parseDictionary = static x => x;
+            }
+        }
+
+        /// <inheritdoc/>
+        public override void Initialize(ISchema schema)
+        {
+            base.Initialize(schema);
+
+            if (_parseDictionary == null) // when typeof(TSourceType) != typeof(object)
+            {
+                // check the value converter for a conversion from dictionary to this object type
+                var conv = ValueConverter.GetConversion(typeof(IDictionary<string, object?>), typeof(TSourceType));
+                if (conv != null)
+                {
+                    _parseDictionary = conv;
+                }
+                else if (GlobalSwitches.DynamicallyCompileToObject)
+                {
+                    // validate and compile the conversion from dictionary to object
+                    _parseDictionary = ObjectExtensions.CompileToObject(typeof(TSourceType), this);
```

Comment:
So the user now must override the `Initialize` method when overriding `ParseDictionary` to avoid the `CompileToObject` penalty? Fills smelly...

Reply:
Maybe `_parseDictionary` can be changed to `Lazy<>`? However, this will take us a step back in the startup validation story.

Reply:
No I can do that because I split ReflectionInfo from the compilation. So long as it can gather the reflection info then it can lazy compile.  I forgot about that scenario. Note that the user will need to override Initialize if they are overriding ParseDictionary for a type that fails our validation rules.

Reply:
It just adds a couple more allocations per input type.

Reply:
Take a look now.  So now it examines the type to determine if ParseDictionary has been overridden.  Solves multiple problems at once; now nobody needs to override Initialize when the override ParseDictionary.  Test added.  Still compiles ahead of time when ParseDictionary has not been overridden.

Reply:
Nice trick!